File: packages/detector-core/aiprojectdetector/scoring/ensemble.py
"""Combine detector signals into a calibrated, explainable AIScore."""

from __future__ import annotations

from ..models import AIScore, Classification, Signal
from ..utils.text import clamp, logistic, mean, stdev
from .calibration import DEFAULT_PROFILE, CalibrationProfile


def classify(ai_probability: float, confidence: float, profile: CalibrationProfile) -> Classification:
    """Map (probability, confidence) to a hedged classification bucket."""
    if confidence < profile.min_confidence:
        return Classification.uncertain
    if ai_probability >= profile.t_generated:
        return Classification.likely_ai_generated
    if ai_probability >= profile.t_assisted_high:
        return Classification.likely_ai_assisted
    if ai_probability >= profile.t_assisted_low:
        return Classification.possibly_ai_assisted
    if ai_probability <= profile.t_human:
        return Classification.likely_human
    return Classification.uncertain


def combine_signals(
    signals: list[Signal],
    profile: CalibrationProfile | None = None,
    *,
    max_reasons: int = 6,
) -> AIScore:
    """Aggregate signals into an AIScore.

    Aggregation is a confidence-and-weight-weighted average of each signal's
    deviation from neutral (0.5). Confidence reflects both the total amount of
    evidence and how much the signals *agree*; when signals disagree we report
    lower confidence rather than a falsely precise number.
    """
    profile = profile or DEFAULT_PROFILE
    informative = [s for s in signals if s.informative]
    if not informative:
        return AIScore(
            ai_probability=0.5,
            human_probability=0.5,
            confidence=0.0,
            evidence_score=0.0,
            risk_score=0.0,
            classification=Classification.uncertain,
            reasons=["Not enough signal to form an estimate."],
        )

    eff_weights: list[float] = []
    deviations: list[float] = []
    for s in informative:
        mult = profile.weight_overrides.get(s.name, 1.0)
        ew = s.weight * s.confidence * mult
        eff_weights.append(ew)
        deviations.append(s.score - 0.5)

    total_w = sum(eff_weights) or 1e-9
    weighted_dev = sum(w * d for w, d in zip(eff_weights, deviations, strict=False)) / total_w
    ai_probability = clamp(0.5 + weighted_dev * profile.gain)
    human_probability = clamp(1.0 - ai_probability)

    # Confidence: more total weight -> higher; more disagreement -> lower.
    volume = logistic(total_w, k=0.6, x0=4.0)  # ~0.5 at total_w=4
    agreement = 1.0 - clamp(stdev(deviations) / 0.3) if len(deviations) > 1 else 0.6
    confidence = clamp(volume * (0.5 + 0.5 * agreement))

    # Evidence score: amount + strength of concrete evidence items.
    ev_items = [e for s in informative for e in s.evidence]
    ev_strength = mean([e.severity for e in ev_items]) if ev_items else 0.0
    evidence_score = clamp(logistic(len(ev_items), k=0.5, x0=3.0) * (0.4 + 0.6 * ev_strength))

    risk_score = clamp(ai_probability * confidence)
    classification = classify(ai_probability, confidence, profile)
    reasons = _top_reasons(informative, eff_weights, deviations, max_reasons)

    return AIScore(
        ai_probability=round(ai_probability, 4),
        human_probability=round(human_probability, 4),
        confidence=round(confidence, 4),
        evidence_score=round(evidence_score, 4),
        risk_score=round(risk_score, 4),
        classification=classification,
        reasons=reasons,
    )


def _top_reasons(
    signals: list[Signal],
    eff_weights: list[float],
    deviations: list[float],
    limit: int,
) -> list[str]:
    ranked = sorted(
        zip(signals, eff_weights, deviations, strict=False),
        key=lambda t: t[1] * abs(t[2]),
        reverse=True,
    )
    reasons: list[str] = []
    for s, _w, dev in ranked:
        if not s.reason:
            continue
        lean = "leans AI" if dev > 0 else "leans human"
        reasons.append(f"[{s.name}, {lean}] {s.reason}")
        if len(reasons) >= limit:
            break
    return reasons

File: packages/detector-core-ts/src/scoring/ensemble.ts
/** Combine detector signals into a calibrated, explainable AIScore. */

import { isInformative, makeAIScore, type AIScore, type Classification, type Signal } from "../models.js";
import { clamp, logistic, mean, round, stdev } from "../utils/text.js";
import { DEFAULT_PROFILE, type CalibrationProfile } from "./calibration.js";

/** Map (probability, confidence) to a hedged classification bucket. */
export function classify(
  aiProbability: number,
  confidence: number,
  profile: CalibrationProfile,
): Classification {
  if (confidence < profile.min_confidence) return "uncertain";
  if (aiProbability >= profile.t_generated) return "likely_ai_generated";
  if (aiProbability >= profile.t_assisted_high) return "likely_ai_assisted";
  if (aiProbability >= profile.t_assisted_low) return "possibly_ai_assisted";
  if (aiProbability <= profile.t_human) return "likely_human";
  return "uncertain";
}

export interface CombineOptions {
  maxReasons?: number;
}

/**
 * Aggregate signals into an AIScore. Aggregation is a confidence-and-weight
 * weighted average of each signal's deviation from neutral (0.5). Confidence
 * reflects both the total amount of evidence and how much signals agree.
 */
export function combineSignals(
  signals: Signal[],
  profile: CalibrationProfile = DEFAULT_PROFILE,
  options: CombineOptions = {},
): AIScore {
  const maxReasons = options.maxReasons ?? 6;
  const informative = signals.filter(isInformative);
  if (informative.length === 0) {
    return makeAIScore({
      ai_probability: 0.5,
      human_probability: 0.5,
      confidence: 0.0,
      evidence_score: 0.0,
      risk_score: 0.0,
      classification: "uncertain",
      reasons: ["Not enough signal to form an estimate."],
    });
  }

  const effWeights: number[] = [];
  const deviations: number[] = [];
  for (const s of informative) {
    const mult = profile.weight_overrides[s.name] ?? 1.0;
    effWeights.push(s.weight * s.confidence * mult);
    deviations.push(s.score - 0.5);
  }

  const totalW = effWeights.reduce((a, b) => a + b, 0) || 1e-9;
  let weightedDev = 0;
  for (let i = 0; i < effWeights.length; i++) weightedDev += effWeights[i]! * deviations[i]!;
  weightedDev /= totalW;
  const aiProbability = clamp(0.5 + weightedDev * profile.gain);
  const humanProbability = clamp(1.0 - aiProbability);

  // Confidence: more total weight -> higher; more disagreement -> lower.
  const volume = logistic(totalW, 0.6, 4.0); // ~0.5 at totalW=4
  const agreement = deviations.length > 1 ? 1.0 - clamp(stdev(deviations) / 0.3) : 0.6;
  const confidence = clamp(volume * (0.5 + 0.5 * agreement));

  // Evidence score: amount + strength of concrete evidence items.
  const evItems = informative.flatMap((s) => s.evidence);
  const evStrength = evItems.length ? mean(evItems.map((e) => e.severity)) : 0.0;
  const evidenceScore = clamp(logistic(evItems.length, 0.5, 3.0) * (0.4 + 0.6 * evStrength));

  const riskScore = clamp(aiProbability * confidence);
  const classification = classify(aiProbability, confidence, profile);
  const reasons = topReasons(informative, effWeights, deviations, maxReasons);

  return makeAIScore({
    ai_probability: round(aiProbability, 4),
    human_probability: round(humanProbability, 4),
    confidence: round(confidence, 4),
    evidence_score: round(evidenceScore, 4),
    risk_score: round(riskScore, 4),
    classification,
    reasons,
  });
}

function topReasons(
  signals: Signal[],
  effWeights: number[],
  deviations: number[],
  limit: number,
): string[] {
  const ranked = signals
    .map((s, i) => ({ s, w: effWeights[i]!, dev: deviations[i]! }))
    .sort((a, b) => b.w * Math.abs(b.dev) - a.w * Math.abs(a.dev));
  const reasons: string[] = [];
  for (const { s, dev } of ranked) {
    if (!s.reason) continue;
    const lean = dev > 0 ? "leans AI" : "leans human";
    reasons.push(`[${s.name}, ${lean}] ${s.reason}`);
    if (reasons.length >= limit) break;
  }
  return reasons;
}
